compiler/rustc_target/src/spec/maybe_lazy.rs

```diff
@@ -0,0 +1,144 @@
+//! A custom LazyLock+Cow suitable for holding borrowed, owned or lazy data.
+
+use std::borrow::{Borrow, Cow};
+use std::fmt::{Debug, Display};
+use std::ops::Deref;
+use std::sync::LazyLock;
+
+enum MaybeLazyInner<T: 'static + ToOwned + ?Sized, F> {
+    Lazy(LazyLock<T::Owned, F>),
+    Cow(Cow<'static, T>),
+}
+
+/// A custom LazyLock+Cow suitable for holding borrowed, owned or lazy data.
+///
+/// Technically this structure has 3 states: borrowed, owned and lazy
+/// They can all be constructed from the [`MaybeLazy::borrowed`], [`MaybeLazy::owned`] and
+/// [`MaybeLazy::lazy`] methods.
+#[repr(transparent)]
+pub struct MaybeLazy<T: 'static + ToOwned + ?Sized, F = fn() -> <T as ToOwned>::Owned> {
+    // Inner state.
+    //
+    // Not to be inlined since we may want in the future to
+    // make this struct usable to statics and we might need to
+    // workaround const-eval limitation (particulary around drop).
+    inner: MaybeLazyInner<T, F>,
+}
+
+impl<T: 'static + ?Sized + ToOwned, F: FnOnce() -> T::Owned> MaybeLazy<T, F> {
```

Comment:
Technicall the `FnOnce` bound is unnecessary for the `borrowed` and `owned` methods.

Its removal should make some other bounds below, e.g. in `From<&'static str>`, unnecessary.

Reply:
I think it is, we are using `MaybeLazy::owned` with a `FnOnce`:
```
error[E0308]: mismatched types
    --> compiler/rustc_target/src/spec/mod.rs:3101:38
     |
2768 | /         macro_rules! key {
2769 | |             ($key_name:ident) => ( {
2770 | |                 let name = (stringify!($key_name)).replace("_", "-");
2771 | |                 if let Some(s) = obj.remove(&name).and_then(|s| s.as_str().map(str::to_strin...
...    |
3101 | |                     base.$key_name = MaybeLazy::owned(args);
     | |                     --------------   ^^^^^^^^^^^^^^^^^^^^^^ expected `MaybeLazy<BTreeMap<..., ...>, ...>`, found `MaybeLazy<_, fn() -> _>`
     | |                     |
     | |                     expected due to the type of this binding
...    |
3172 | |             } );
3173 | |         }
     | |_________- in this expansion of `key!`
...
3205 |           key!(post_link_objects = "post-link-objects", link_objects);
     |           ----------------------------------------------------------- in this macro invocation
     |
     = note: expected struct `MaybeLazy<BTreeMap<LinkOutputKind, Vec<Cow<'static, str>>>, LazyCrtObjectsState>`
                found struct `MaybeLazy<_, fn() -> _>`
```